parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,659 @@
+    // Similar to the `readLong` loop, but all bytes should be unsigned-extended.
+    for (int i = 0; i < numBytes; ++i) {
+      int unsignedByteValue = bytes.get(pos + i) & 0xFF;
+      result |= unsignedByteValue << (8 * i);
+    }
+    if (result < 0) {
+      throw new IllegalArgumentException(String.format("Failed to read unsigned int. numBytes: %d", numBytes));
+    }
+    return result;
+  }
+
+  /**
+   * Returns the value type of Variant value `value[pos...]`. It is only legal to call `get*` if
+   * `getType` returns the corresponding type. For example, it is only legal to call
+   * `getLong` if this method returns `Type.Long`.
+   * @param value The Variant value to get the type from
+   * @return The type of the Variant value
+   */
+  static Variant.Type getType(ByteBuffer value) {
+    checkIndex(value.position(), value.limit());
+    int basicType = value.get(value.position()) & BASIC_TYPE_MASK;
+    int typeInfo = (value.get(value.position()) >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    switch (basicType) {
+      case SHORT_STR:
+        return Variant.Type.STRING;
+      case OBJECT:
+        return Variant.Type.OBJECT;
+      case ARRAY:
+        return Variant.Type.ARRAY;
+      default:
+        switch (typeInfo) {
+          case NULL:
+            return Variant.Type.NULL;
+          case TRUE:
+          case FALSE:
+            return Variant.Type.BOOLEAN;
+          case INT8:
+            return Variant.Type.BYTE;
+          case INT16:
+            return Variant.Type.SHORT;
+          case INT32:
+            return Variant.Type.INT;
+          case INT64:
+            return Variant.Type.LONG;
+          case DOUBLE:
+            return Variant.Type.DOUBLE;
+          case DECIMAL4:
+            return Variant.Type.DECIMAL4;
+          case DECIMAL8:
+            return Variant.Type.DECIMAL8;
+          case DECIMAL16:
+            return Variant.Type.DECIMAL16;
+          case DATE:
+            return Variant.Type.DATE;
+          case TIMESTAMP_TZ:
+            return Variant.Type.TIMESTAMP_TZ;
+          case TIMESTAMP_NTZ:
+            return Variant.Type.TIMESTAMP_NTZ;
+          case FLOAT:
+            return Variant.Type.FLOAT;
+          case BINARY:
+            return Variant.Type.BINARY;
+          case LONG_STR:
+            return Variant.Type.STRING;
+          case TIME:
+            return Variant.Type.TIME;
+          case TIMESTAMP_NANOS:
+            return Variant.Type.TIMESTAMP_NANOS;
+          case TIMESTAMP_NANOS_NTZ:
+            return Variant.Type.TIMESTAMP_NANOS_NTZ;
+          case UUID:
+            return Variant.Type.UUID;
+          default:
+            throw new UnsupportedOperationException(
+                String.format("Unknown type in Variant. primitive type: %d", typeInfo));
+        }
+    }
+  }
+
+  private static IllegalArgumentException unexpectedType(Variant.Type type) {
+    return new IllegalArgumentException("Expected type to be " + type);
```

Comment:
I think this may be phrased incorrectly. Isn't the problem that a method like `getUUID` was called for a value that is not a UUID? In that case do we trust the user called the right method and the expectation is valid? Instead, I would say something more direct that doesn't make that assumption: `"Cannot read %s value as a %s", actualType, expectedType`